File: solutions/131/solution.rb
class String
  def is_palindrome?(lo, hi)
    while lo < hi
      return false if self[lo] != self[hi]
      lo += 1
      hi -= 1
    end
    true
  end
end

# @param {String} s
# @return {String[][]}
def partition(s)
  res = []
  
  cur = []
  traverse = lambda do |lo|
    if lo == s.size
      res << cur.dup
      return
    end

    (lo...s.size).each do |hi|
      next unless s.is_palindrome?(lo, hi)
      cur << s[lo..hi]
      traverse.(hi + 1)
      cur.pop
    end
  end

  traverse.(0)

  res
end

File: solutions/131/solution.js
function isPalindrome(s, lo, hi) {
  while (lo < hi) {
    if (s[lo++] !== s[hi--]) {
      return false;
    }
  }
  return true;
}

/**
 * @param {string} s
 * @return {string[][]}
 */
function partition(s) {
  const res = [];

  const cur = [];
  const traverse = (start) => {
    if (start === s.length) {
      res.push([...cur]);
      return;
    }

    for (let end = start; end < s.length; ++end) {
      if (!isPalindrome(s, start, end)) {
        continue;
      }

      cur.push(s.slice(start, end + 1));
      traverse(end + 1);
      cur.pop();
    }
  };

  traverse(0);

  return res;
}
